HoRNDIS.cpp

```diff
@@ -28,516 +30,343 @@
+	if (interface == NULL) {
+		LOG(V_ERROR, "unexpected provider class (wrong Info.plist)");
 		return NULL;
 	}
 	
-	/* And look for one of the options. */
-	IOUSBInterfaceDescriptor *descout;
-	IOUSBFindInterfaceRequest req;
-	bool found = false;
-	
-	ctrlconfig = cd->bConfigurationValue;
-	
-	req.bInterfaceClass = 2;
-	req.bInterfaceSubClass = 2;
-	req.bInterfaceProtocol = 255;
-	if (dev->FindNextInterfaceDescriptor(cd, NULL, &req, &descout) == kIOReturnSuccess) {
-		LOG(V_NOTE, "probe: looks like we're good (2/2/255)");
-		ctrlclass = 2;
-		ctrlsubclass = 2;
-		ctrlprotocol = 255;
-		found = true;
+	if (!isRNDISControlInterface(interface->getInterfaceDescriptor())) {
+		LOG(V_ERROR, "not RNDIS control interface (wrong Info.plist)");
+		return NULL;
 	}
-
-	req.bInterfaceClass = 224;
-	req.bInterfaceSubClass = 1;
-	req.bInterfaceProtocol = 3;
-	if (dev->FindNextInterfaceDescriptor(cd, NULL, &req, &descout) == kIOReturnSuccess) {
-		ctrlclass = 224;
-		ctrlsubclass = 1;
-		ctrlprotocol = 3;
-		LOG(V_NOTE, "probe: looks like we're good (224/1/3)");
-		found = true;
+	const uint8_t controlIfNum = interface->getInterfaceDescriptor()->bInterfaceNumber;
+	
+	// Now, search for data interface: if found, we're done:
+	bool foundData = false;
+	const ConfigurationDescriptor *confDesc = interface->getConfigurationDescriptor();
+	const InterfaceDescriptor* intDesc = NULL;
+	while((intDesc = StandardUSB::getNextInterfaceDescriptor(confDesc, intDesc)) != NULL) {
+		// Just in case an Android device has several CDC data interfaces, we would only
+		// pick the one that follows RIGHT AFTER the RNDIS interface:
+		if (isCDCDataInterface(intDesc) &&
+			intDesc->bInterfaceNumber == controlIfNum + 1) {
+			foundData = true;
+			break;
+		}
 	}
 	
-	if (!found) {
-		LOG(V_NOTE, "probe: this composite device is not for us");
+	if (foundData) {
+		*score += 100000;
+		return this;
+	} else {
+		// Did not find any interfaces we can use:
+		LOG(V_DEBUG, "unexpected provider class or parameters: this device is not for us");
 		return NULL;
 	}
-	
-	*score += 10000;
-	return this;
 }
 
 /***** Ethernet interface bits *****/
 
-/* We need our own createInterface (overriding the one in IOEthernetController) because we need our own subclass of IOEthernetInterface.  Why's that, you say?  Well, we need that because that's the only way to set a different default MTU.  Sigh... */
+/* We need our own createInterface (overriding the one in IOEthernetController) 
+ * because we need our own subclass of IOEthernetInterface.  Why's that, you say?
+ * Well, we need that because that's the only way to set a different default MTU,
+ * because it seems like MacOS code just assumes that any EthernetController
+ * driver must be able to handle al leaset 1500-byte Ethernet payload.
+ * Sigh...
+ * The MTU-limiting code may never come into play though, because the devices
+ * I've seen have "max_transfer_size" large enough to accomodate a max-length 
+ * Ethernet frames. */
```

Comment:
I think there are weird embedded devices that have less than that.

Reply:
That's why I decided to leave this logic in-place.